src/notify/notify.c

```diff
@@ -313,56 +337,227 @@ static int parse_argv(int argc, char *argv[]) {
+        assert(fd >= 0);
+
+        _cleanup_free_ char *text = NULL;
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        r = notify_recv(fd, &text, /* ret_ucred= */ NULL, &pidref);
+        if (r == -EAGAIN)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to receive notification message: %m");
+
+        if (!pidref_equal(child, &pidref)) {
+                log_warning("Received notification message from unexpected process " PID_FMT " (expected " PID_FMT "), ignoring.",
+                            pidref.pid, child->pid);
+                return 0;
         }
 
-        if (!fdset_isempty(passed))
-                log_warning("Warning: %u more file descriptors passed than referenced with --fd=.", fdset_size(passed));
+        const char *p = find_line_startswith(text, "READY=1");
+        if (!p || !IN_SET(*p, '\n', 0)) {
+                if (DEBUG_LOGGING) {
+                        _cleanup_free_ char *escaped = cescape(text);
+                        log_debug("Received notification message without READY=1, ignoring. (%s)", strna(escaped));
+                }
+        } else {
+                log_debug("Recieved READY=1, exiting.");
 
-        return 1;
+                r = sd_event_exit(sd_event_source_get_event(s), EXIT_SUCCESS);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to request event loop exit: %m");
+        }
+
+        return 0;
+}
+
+static int on_child(sd_event_source *s, const siginfo_t *si, void *userdata) {
+        int r;
+
+        assert(s);
+        assert(si);
+
+        int ret;
+        if (si->si_code == CLD_EXITED) {
+                if (si->si_status != EXIT_SUCCESS)
+                        log_debug("Child failed with exit status %i.", si->si_status);
+                else
+                        log_debug("Child exited successfully");
+
+                ret = si->si_code;
+
+        } else if (IN_SET(si->si_code, CLD_KILLED, CLD_DUMPED))
+                ret = log_debug_errno(SYNTHETIC_ERRNO(EPROTO),
+                                       "Child terminated by signal %s.", signal_to_string(si->si_status));
+        else
+                ret = log_debug_errno(SYNTHETIC_ERRNO(EPROTO),
+                                      "Child terminated due to unknown reason.");
+
+        r = sd_event_exit(sd_event_source_get_event(s), ret);
+        if (r < 0)
+                return log_error_errno(r, "Failed to request event loop exit: %m");
+
+        return 0;
+}
+
+static int action_fork(char **_command) {
+        int r;
+
+        assert(!strv_isempty(_command));
+
+        /* Make a copy, since pidref_safe_fork_full() will change argv[] further down. */
+        _cleanup_strv_free_ char **command = strv_copy(_command);
+        if (!command)
+                return log_oom();
+
+        _cleanup_close_ int socket_fd = RET_NERRNO(socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0));
```

Comment:
Is there a point in `RET_ERRNO()` here? Might as well just pass `errno` to `log_error_errno()` no?